// infrastructure/infrastructure.js
// Maybe require('./infrastructure')
'use strict';

function Infrastructure () {
    this.services = {};
    this.features = {};
    var global = null;
    var root = null;

    this.init = function (window, rootElement, options) {
        global = window;
        root = rootElement;

        this.startFeature('shouter');
    };

    this.getGlobal = function (name) {
        if (name in global) { return global[name]; }

        // default return null
        return null;
    };

    this.getRoot = function () {
        return root;
    };
}

Infrastructure.prototype.useService = function (name, factory, options) {
    // separate services object
    // this.services.add(factory, options)
    // Throws custom error
    this.services[name] = {
        factory: factory,
        options: options,
        instance: null
    };
};

Infrastructure.prototype.addFeature = function (name, factory, options) {
    // separate features object
    // this.features.add(factory, options)
    // Throws custom error
    this.features[name] = {
        factory: factory,
        options: options,
        // instance: null
    };
};

Infrastructure.prototype.startFeature = function (name) {
    var feature = this.features[name];
    var element = this.getRoot().querySelector('[data-feature~='+name+']');
    feature.factory(element, this);
};

Infrastructure.prototype.getService = function (name) {
    // separate services object
    // fetch for now retrieve for later
    // this.services.get(factory, options)
    // Throws custom error
    var service = this.services[name],
        instance = service.instance,
        application = this;

    if (instance) { return instance; }

    instance = service.factory(application, service.options);
    service.instance = instance;

    return instance;
};

// Infrastructure.prototype.

module.exports = function () {
    return new Infrastructure();
};
